### PR TITLE
BCI overlays and click interception

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Adds BCI overlay, bar overlay, click interceptor and counter overlay circuits. Moves BCIs and all of it's stuff into a separate research node that requires biotech and costs 500(because I didn't want to stuff all of the BCI-exclusive stuff into "advanced shells" node). Fixes a bunch of BCI bugs and oddities.
image
image
image

## Why It's Good For The Game

Allows you to make cooler stuff with BCIs

## Changelog
:cl:
expansion: You can make BCI overlay and click interceptor circutry now!
balance: BCI now has it's own research node.
fix: You no longer can remove BCI core and actions from BCIs and insert them into other circutry.
/:cl:

### CODE REVIEW

**Comment by Mothblocks** [comment]
you had some cool pictures of this that i want people to see

**Comment by Mothblocks** [comment]
Can you put these in their own category of "BCI Components" or something so that it's even more obvious in the menu? Eventually I'd want to update it a touch so that we can expand that easier to more required shells, and also for components that would support multiple.

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/obj/item/circuit_component/click_intercept
+	display_name = "Click Intercept"
+	display_desc = "A component that when activated will intercept next user's click and output the object that was clicked. Requires a shell."
+
+	required_shells = list(/obj/item/organ/cyberimp/bci)
+
+	var/datum/port/output/clicked_atom
+
+	var/obj/item/organ/cyberimp/bci/bci
+	var/coodown
```
Comment (Watermelon914) [inline code/modules/wiremod/components/hud/click_intercept.dm:17]: You can use UI notices to indicate this is only usable in BCIs

File: code/modules/wiremod/components/hud/counter_overlay.dm [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+	if(. || !bci)
+		return
+
+	var/mob/living/owner = bci.owner
+
+	if(!owner || !istype(owner) || !owner.client)
+		return
+
+	for(var/image/number in numbers)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/hud/counter_overlay.dm:53]:
```suggestion
	for(var/image/number as anything in numbers)
```

File: code/modules/wiremod/components/hud/counter_overlay.dm [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			number.pixel_y = image_pixel_y.input_value
+
+		numbers.Add(number)
+		owner.client.images |= number
+
+/obj/item/circuit_component/counter_overlay/proc/on_organ_removed(datum/source, mob/living/carbon/owner)
+	SIGNAL_HANDLER
+	owner.client.images.Remove(counter)
+	for(var/image/number in numbers)
+		owner.client.images.Remove(number)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/hud/counter_overlay.dm:83]: be aware of the owner getting a new mob

Comment (Mothblocks) [reply code/modules/wiremod/components/hud/counter_overlay.dm:83]: This is what alternate_appearance is for, by the way. If you want an example, look at the in_love status effect, which had an alternate_appearance I made.

File: code/modules/research/techweb/all_nodes.dm [hunk 4]
```diff
@@ -667,6 +665,22 @@
...
+	description = "Grants access to biocompatable shell designs and components."
+	prereq_ids = list("adv_shells", "biotech")
+	design_ids = list(
+		"bci_implanter",
+		"bci_shell",
+		"comp_bci_action",
+		"comp_object_overlay",
+		"comp_bar_overlay",
+		"comp_click_intercept",
+		"comp_counter_overlay",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:680]: ABC order.

File: code/modules/wiremod/components/hud/bar_overlay.dm [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/**
+ * # Bar Overlay Component
+ *
+ * Basically an advanced verion of object overlay component that shows a horizontal/vertical bar.
+ * Requires a BCI shell.
+ */
+
+#define OBJECT_OVERLAY_LIMIT 10
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/bar_overlay.dm:8]: #undef

File: code/modules/wiremod/components/hud/bar_overlay.dm [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+/obj/item/circuit_component/object_overlay/bar
+	display_name = "Bar Overlay"
+	display_desc = "A component that shows a bar overlay ontop of an object. Requires a BCI shell. Requires a 0-100 number to work propperly."
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/bar_overlay.dm:12]:
```suggestion
	display_desc = "Requires a BCI shell. A component that shows a bar overlay ontop of an object from a range of 0 to 100."
```

File: code/modules/wiremod/components/hud/bar_overlay.dm [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		number_clear = round(number_clear / 10) * 10
+	var/image/cool_overlay = image(icon = 'icons/hud/screen_bci.dmi', loc = target_atom, icon_state = "[options_map[current_option]][number_clear]", layer = RIPPLE_LAYER)
+
+	if(image_pixel_x.input_value)
+		cool_overlay.pixel_x = image_pixel_x.input_value
+
+	if(image_pixel_y.input_value)
+		cool_overlay.pixel_y = image_pixel_y.input_value
+
+	var/alt_appearance = WEAKREF(target_atom.add_alt_appearance(
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/bar_overlay.dm:55]: No need for the variable, just `= WEAKREF(` directly.

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 8]
```diff
@@ -0,0 +1,64 @@
+/**
+ * # Click Intercept Component
+ *
+ * When activated intercepts next click and outputs clicked atom.
+ * Requires a BCI shell.
+ */
+
+/obj/item/circuit_component/click_intercept
+	display_name = "Click Intercept"
+	display_desc = "A component that when activated will intercept next user's click and output the object that was clicked. Requires a shell."
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/click_intercept.dm:10]:
```suggestion
	display_desc = "Requires a BCI shell. A component that when activated will intercept next user's click and output the object that was clicked."
```

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 9]
```diff
@@ -0,0 +1,64 @@
+/**
+ * # Click Intercept Component
+ *
+ * When activated intercepts next click and outputs clicked atom.
+ * Requires a BCI shell.
+ */
+
+/obj/item/circuit_component/click_intercept
+	display_name = "Click Intercept"
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/click_intercept.dm:9]: "Click" is fairly OOC, see if you can come up with a better IC term.

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+	var/mob/living/owner = bci.owner
+	if(!owner || !istype(owner) || !owner.client)
+		return
+
+	if(TIMER_COOLDOWN_CHECK(parent, COOLDOWN_CIRCUIT_CLICK_INTERCEPT))
+		return
+
+	to_chat(owner, "<B>Left-click to activate click interceptor!</B>")
+	owner.client.click_intercept = src
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/click_intercept.dm:47]: This is going to be a problem.

Because click interceptors is a *variable*, and not say, a stack, this means that anything that adds its own is going to null this out, requiring it to be reactivated.

For instance--using a spell then cancelling it will not preserve this.

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	required_shells = list(/obj/item/organ/cyberimp/bci)
+
+	var/datum/port/output/clicked_atom
+
+	var/obj/item/organ/cyberimp/bci/bci
+	var/intercept_cooldown = 1 SECONDS
+
+/obj/item/circuit_component/click_intercept/Initialize()
+	. = ..()
+	trigger_input = add_input_port("Activate", PORT_TYPE_SIGNAL)
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/click_intercept.dm:21]: There doesn't appear to be a way to deactivate this? Or at least, the name doesn't imply as such.

File: code/modules/wiremod/components/hud/counter_overlay.dm [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+	for(var/i = 1 to 3)
+		var/cur_num = round(cleared_number / (10 ** (3 - i))) % 10
+		var/image/number = image(icon = 'icons/hud/screen_bci.dmi', icon_state = "hud_number_[cur_num]", loc = owner)
+
+		if(image_pixel_x.input_value)
+			number.pixel_x = image_pixel_x.input_value + (i - 1) * 9
+		if(image_pixel_y.input_value)
+			number.pixel_y = image_pixel_y.input_value
+
+		var/number_appearance = WEAKREF(owner.add_alt_appearance(
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/counter_overlay.dm:88]: No need to make a variable.

File: code/modules/wiremod/components/hud/counter_overlay.dm [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+	var/datum/port/input/image_pixel_x
+	var/datum/port/input/image_pixel_y
+
+	var/datum/port/input/signal_update
+
+	var/obj/item/organ/cyberimp/bci/bci
+	var/image/counter
+	var/list/numbers = list()
+	var/counter_appearance
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/counter_overlay.dm:24]: Make sure to destroy this.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 14]
```diff
@@ -36,10 +36,12 @@
 /obj/item/circuit_component/bci
 	display_name = "Brain-Computer Interface"
 	display_desc = "Used to receive inputs for the brain-computer interface. User is presented with three buttons."
+	removable = FALSE
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:39]: I don't think this is actually used, just remove it. I was probably trying to do something else.

File: code/modules/wiremod/components/hud/object_overlay.dm [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+	var/datum/port/input/image_pixel_x
+	var/datum/port/input/image_pixel_y
+
+	/// On/Off signals
+	var/datum/port/input/signal_on
+	var/datum/port/input/signal_off
+
+	var/obj/item/organ/cyberimp/bci/bci
+	var/list/active_overlays = list()
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/object_overlay.dm:27]: Make sure to qdel these.

File: code/modules/wiremod/components/hud/object_overlay.dm [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+	var/list/atom_locs = list()
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/object_overlay.dm:28]: This is unused.

File: code/modules/wiremod/components/hud/object_overlay.dm [hunk 17]
```diff
@@ -0,0 +1,132 @@
+/**
+ * # Object Overlay Component
+ *
+ * Shows an overlay ontop of an object. Toggleable.
+ * Requires a BCI shell.
+ */
+
+GLOBAL_VAR_INIT(object_overlay_id, 0) //I need every object_overlay component to have a DIFFERENT ID because else they are going to overlap
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/object_overlay.dm:8]: This needs a doc comment as two places used it, and should be moved up since it is used outside of object_overlays.

File: code/modules/wiremod/components/hud/object_overlay.dm [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+
+/obj/item/circuit_component/object_overlay
+	display_name = "Object Overlay"
+	display_desc = "A component that shows an overlay ontop of an object. Requires a BCI shell."
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/object_overlay.dm:12]:
```suggestion
	display_desc = "Requires a BCI shell. A component that shows an overlay on top of an object."
```

File: code/modules/wiremod/components/hud/bar_overlay.dm [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+/obj/item/circuit_component/object_overlay/bar
+	display_name = "Bar Overlay"
+	display_desc = "A component that shows a bar overlay ontop of an object. Requires a BCI shell. Requires a 0-100 number to work propperly."
+
+	var/datum/port/input/bar_number
+
+/obj/item/circuit_component/object_overlay/bar/Initialize()
+	. = ..()
+	bar_number = add_input_port("Number", PORT_TYPE_ATOM)
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/bar_overlay.dm:18]: Number, port type atom?

File: code/modules/research/designs/wiremod_designs.dm [hunk 20]
```diff
@@ -238,6 +238,33 @@
...
+/datum/design/component/bci
+	category = list("Circuitry", "BCI Components")
+
+/datum/design/component/bci/bci_action
+	name = "BCI Action Component"
+	id = "comp_bci_action"
+	build_path = /obj/item/circuit_component/bci_action
+
+/datum/design/component/bci/object_overlay
+	name = "BCI Action Component"
```
Comment (Mothblocks) [inline code/modules/research/designs/wiremod_designs.dm:250]: Wrong name

File: code/modules/research/techweb/all_nodes.dm [hunk 21]
```diff
@@ -667,6 +665,22 @@
 	)
 	research_costs = list(TECHWEB_POINT_TYPE_GENERIC = 2500)
 
+/datum/techweb_node/bci_shells
+	id = "bci_shells"
+	display_name = "Brain-Computer Interfaces"
+	description = "Grants access to biocompatable shell designs and components."
+	prereq_ids = list("adv_shells", "biotech")
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:672]: -1 to requiring biotech, at least not this early.

```suggestion
	prereq_ids = list("adv_shells")
```

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 22]
```diff
@@ -157,6 +159,8 @@
 	display_name = "BCI Core"
 	display_desc = "Controls the core operations of the BCI."
 
+	removable = FALSE
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:162]: This isn't necessary. It being a part of the shell *already* makes it non-removable. This just makes it less generic.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/wiremod/components/hud/bar_overlay.dm (hunk 5 above)
Comment (SmArtKar) [reply code/modules/wiremod/components/hud/bar_overlay.dm:8]: its kinda used in object_overlay.dm

File: code/modules/wiremod/components/hud/click_intercept.dm (hunk 9 above)
Comment (SmArtKar) [reply code/modules/wiremod/components/hud/click_intercept.dm:9]: I thought of point instead of click but we also have "point" verb soooo

File: code/modules/wiremod/components/hud/click_intercept.dm (hunk 10 above)
Comment (SmArtKar) [reply code/modules/wiremod/components/hud/click_intercept.dm:47]: Yeah I know but it's not really going to break it.

File: code/modules/wiremod/components/hud/click_intercept.dm (hunk 11 above)
Comment (SmArtKar) [reply code/modules/wiremod/components/hud/click_intercept.dm:21]: You wait until user clicks on something

File: code/modules/wiremod/components/hud/bar_overlay.dm (hunk 5 above)
Comment (SmArtKar) [reply code/modules/wiremod/components/hud/bar_overlay.dm:8]: but okay

File: code/modules/wiremod/components/hud/bar_overlay.dm [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+	display_name = "Bar Overlay"
+	display_desc = "Requires a BCI shell. A component that shows a bar overlay ontop of an object from a range of 0 to 100."
+
+	var/datum/port/input/bar_number
+
+/obj/item/circuit_component/object_overlay/bar/Initialize()
+	. = ..()
+	bar_number = add_input_port("Number", PORT_TYPE_NUMBER)
+
+	set_option("Vertical")
```
Comment (Watermelon914) [inline code/modules/wiremod/components/hud/bar_overlay.dm:20]: If you want vertical to be the default, have it first in the component_options_bar list.

File: code/modules/wiremod/components/hud/object_overlay.dm [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+	. = ..()
+	target = add_input_port("Target", PORT_TYPE_ATOM)
+
+	signal_on = add_input_port("Create Overlay", PORT_TYPE_SIGNAL)
+	signal_off = add_input_port("Remove Overlay", PORT_TYPE_SIGNAL)
+
+	image_pixel_x = add_input_port("X-Axis Shift", PORT_TYPE_NUMBER)
+	image_pixel_y = add_input_port("Y-Axis Shift", PORT_TYPE_NUMBER)
+
+	set_option("Corners (Blue)")
```
Comment (Watermelon914) [inline code/modules/wiremod/components/hud/object_overlay.dm:42]: By default this'll already be the first option

**Comment by SmArtKar** [comment]
Actually I am a retard and instead of that global counter shit I could've used REF(src) what I successfully did

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Me and Watermelon talked--name it Target Intercept or similar, click is OOC.

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 25]
```diff
@@ -19,8 +19,8 @@
 /obj/item/circuit_component/click_intercept/Initialize()
 	. = ..()
 	trigger_input = add_input_port("Activate", PORT_TYPE_SIGNAL)
-	trigger_output = add_output_port("Clicked", PORT_TYPE_SIGNAL)
-	clicked_atom = add_output_port("Clicked Object", PORT_TYPE_ATOM)
+	trigger_output = add_output_port("Target", PORT_TYPE_SIGNAL)
```
Comment (Watermelon914) [inline code/modules/wiremod/components/hud/click_intercept.dm:22]: Call it triggered or something, to indicate it's a signal. Target suggests PORT_TYPE_ATOM

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 26]
```diff
@@ -0,0 +1,64 @@
+/**
+ * # Click Intercept Component
+ *
+ * When activated intercepts next click and outputs clicked atom.
+ * Requires a BCI shell.
+ */
+
+/obj/item/circuit_component/click_intercept
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/click_intercept.dm:8]: Rename the path and comments and everything to target_intercept as well to remove confusion.

File: code/modules/wiremod/components/hud/click_intercept.dm [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+	display_name = "Target Intercept"
+	desc = "Requires a BCI shell. A component that when activated will intercept next user's click and output the object that was clicked."
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/click_intercept.dm:10]: Reword this to use target instead of click.

File: code/modules/wiremod/components/hud/counter_overlay.dm [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+	SIGNAL_HANDLER
+	QDEL_NULL(counter_appearance)
+	for(var/number in numbers)
+		QDEL_NULL(number)
+
+/obj/item/circuit_component/counter_overlay/Destroy()
+	QDEL_NULL(counter_appearance)
+	for(var/number in numbers)
+		QDEL_NULL(number)
+	. = ..()
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/counter_overlay.dm:102]:
```suggestion
	return ..()
```

File: code/modules/wiremod/components/hud/object_overlay.dm [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+	signal_on = add_input_port("Create Overlay", PORT_TYPE_SIGNAL)
+	signal_off = add_input_port("Remove Overlay", PORT_TYPE_SIGNAL)
+
+	image_pixel_x = add_input_port("X-Axis Shift", PORT_TYPE_NUMBER)
+	image_pixel_y = add_input_port("Y-Axis Shift", PORT_TYPE_NUMBER)
+
+/obj/item/circuit_component/object_overlay/Destroy()
+	for(var/active_overlay in active_overlays)
+		QDEL_NULL(active_overlay)
+	. = ..()
```
Comment (Mothblocks) [inline code/modules/wiremod/components/hud/object_overlay.dm:43]:
```suggestion
	return ..()
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Last few requests.
